File: Java/PathSum.java
// iterative solution that uses two stacks sum to keep track of everything
// O(n) time complexity

import java.util.Stack;
class Solution {
    public boolean hasPathSum(TreeNode root, int targetSum) {
        if (root == null) return false;
        Stack<TreeNode> nodeStack = new Stack<>();
        Stack<Integer> numStack = new Stack<>();
        nodeStack.add(root);
        numStack.add(root.val);

        while (!nodeStack.isEmpty()) {
            TreeNode node = nodeStack.pop();
            int num = numStack.pop();
            
            if (node.left == null && node.right == null) if (num == targetSum) return true;
            
            if (node.right != null) {
                nodeStack.add(node.right);
                numStack.add(node.right.val + num);
            }
            if (node.left != null) {
                nodeStack.add(node.left);
                numStack.add(node.left.val + num);
            }
        }
        return false;
    }
}



// recursive solution 
// O(n) dfs
class Solution {
    public boolean hasPathSum(TreeNode root, int targetSum) {
        return dfs(root, 0, targetSum);
    }

    public boolean dfs (TreeNode node, int currSum, int targetSum) {
        if (node == null) return false;
        currSum += node.val;
        if (node.left == null && node.right == null) {
            return currSum == targetSum;
        }

        return dfs(node.right, currSum, targetSum) || dfs(node.left, currSum, targetSum);
        
    }

}

File: Python/Backtracking/pathSum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from typing import Optional


class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool: # type: ignore
        

        def backtrack(node, currSum):
            #leaf node
            if not node.left and not node.right:
                if currSum == targetSum:
                    return True

            if node.left:
                if backtrack(node.left, currSum + node.left.val):
                    return True
            if node.right:
                if backtrack(node.right, currSum + node.right.val):
                    return True

            return False

            
        if root:
            return backtrack(root, root.val)
        else:
            return False